### ISSUE TITLE
Add Worktrees to the left pane

### ISSUE THREAD

**Issue by KvanTTT** [gitextensions/gitextensions#7048]
## Feature description

The idea is that this would effectively replace the Worktrees menu, and perhaps even the Worktrees form, combining all operations in one place. The same as #5569 but for worktrees. I think it will be very useful for some users. See why worktrees are useful on the [official doc](https://git-scm.com/docs/git-worktree) and on [stackoverflow](https://stackoverflow.com/a/31951225/1046374).

It will be possible to switch to a certain worktree just with a double-click. See the image for clarity:

![image](https://user-images.githubusercontent.com/1150330/63654386-45aaad00-c782-11e9-8c4d-80b16f72812d.png)

If worktrees are not used, this item can be hidden.

## Environment

- Git Extensions 3.1.1.6049
- Build 2f872105b4eb42421d9fdafd68e90a9e27d20d96
- Git 2.21.0.windows.1
- Microsoft Windows NT 10.0.17763.0
- .NET Framework 4.8.3801.0
- DPI 96dpi (no scaling)

**Comment by gerhardol** [comment]
You probably want to see both a name (last part of base path?) and branch in the left panel

**Comment by KvanTTT** [comment]
Yes, it's reasonable. And maybe full path on mouseover.

**Comment by pmiossec** [comment]
> The idea is that this would effectively replace the Worktrees menu, and perhaps even the Worktrees form, combining all operations in one place.

My main concern is the performance issue.
I'm not against it because I understand that could be useful for some users (that's why I developed the worktree feature 😉 ) but... we are putting a lot of things in this left panel and we already know that it is slowing down the refresh of the UI.

From what I remember a lot of  git commands are run to retrieve the data to fill the left panel and, this, at each grid revision refresh!

So, for me, the left panel should be kept to display only data that you use often.

And correct me if I am wrong, but we switch occasionally between workspaces. If you really have to work with 2 workspaces in the same time, it's much more efficient to open 2 GitExtensions instances.

I also know that you could configure GitExtensions to hide some things in the left ̂panel (to prevent performance issue) but, if you have to enable and disable it often, that goes against the added value to have the worktrees list always displayed in the left panel.

To sum up my point of view:

* I don't see a great added value
* I expect some performance problem
* If it's done, it should be disabled by default (to avoid the performance cost on a daily use of git)
* that's must not replace the "Worktrees" menu (for the ones that disabled the feature in the left panel)

**A new "Switch to workspace" menu item containing the list of the workspaces in the already  "Worktrees" menu item, to easily switch between workspaces looks to me to be a better idea.**

**Comment by KvanTTT** [comment]
> My main concern is the performance issue.

The number of actual worktrees not so big (usually up to 3). I don't think it affects performance at all.

> And correct me if I am wrong, but we switch occasionally between worktrees . If you really have to work with 2 worktrees in the same time, it's much more efficient to open 2 GitExtensions instances.

Yes, you can open 2 GitE instances but it's not so convenient and clearly as a single instance is. Sometimes I forget what branch checkout on the current directory. It forces me to open a new instance or go to the Worktress menu. The similar concept with submodules: you can use several instances but the implementation of #5569 significantly improved UX. Do submodules affect performance? I often use switching between worktrees.

**Comment by pmiossec** [comment]
>I don't think it affects performance at all.

That adds up to the others and we **already** have a performance problem on the left panel. That's why I'm a little reluctant (but not against it!). And also that avoid clutters by having a lot of things in the left panel.

>Is it your suggestion on how to implement or it already exists?

It is a proposition of alternative to your feature request. When I work inside a repository, I personally don't have to see all the times all the worktrees, just when I want to switch.

But, thinking of it again, I remember why I didn't do much than that we have today: Once opened, the worktree appears in the recent repository list. Isn't it enough for a quick access (if you switch **often** between worktrees)?

**Comment by KvanTTT** [comment]
> Once opened, the worktree appears in the recent repository list. Isn't it enough for a quick access (if you switch often between worktrees)?

Not completely, because of the name of a branch is not shown, only full path. In this case, maybe it makes sense to add a checkout branch name after the full path (`C:\path\to\repostiroty [master]`). Moreover, there is no link between repositories in this list, but worktrees repositories are linked. On the other hand, it's possible to mark worktree repositories with the pictogram:

![image](https://user-images.githubusercontent.com/1150330/63689090-adf49f80-c811-11e9-8480-65e107f39f76.png)

> It is a proposition of alternative to your feature request. When I work inside a repository, I personally don't have to see all the times all the worktrees, just when I want to switch.

Ok, I don't have an objection on such implementation.

**Comment by KvanTTT** [comment]
Yet another thought: group worktrees and submodules in recent repositories list in the bottom part.

**Comment by gerhardol** [comment]
I would like to (optionally) exclude submodules from recent: #4942
(to get relevant lists, I work with many superprojects at the same time).

The worktree path-branch info is only interesting "sporadically".
I find Manage Worktree form a little hard to get to, a button or shortcut may be sufficient

**Comment by RussKie** [comment]
> My main concern is the performance issue.

Thank you for highlighting this often overlooked point.


> So, for me, the left panel should be kept to display only data that you use often.

That is why the left panel offers functionality to show/hide required nodes.
Worktrees can be just another node we surface, and users can show/hide them as required.

In the docs we can highlight the fact that showing all nodes in the left panel may affect the performance of the app.

**Comment by RussKie** [comment]
> That adds up to the others and we **already** have a performance problem on the left panel.

Could you please elaborate?
I am unable to recall any significant reports about it. We had some general perf issues in 3.0, but 3.1.1 isn't suffering from major perf issues AFAIK.

**Comment by gerhardol** [comment]
In 3.2 #6844 is fixed that was reported as a left panel problem.
There are not much extra running for left panel, listing ahead/behind comes to mind.
So I am not aware of any issues.

Listing worktrees will be done async and should not affect performance at all.
It will be useful occasionally, I am still not sure it is the best way to fix the problem.
Right now I considering a button to start worktree form would be a good solution.

**Comment by RussKie** [comment]
So to sum up, we would accept a PR that adds worktrees node to the left panel.
As a quick workaround, you could follow @gerhardol's directions.

**Comment by gerhardol** [comment]
An alternative could be to add worktrees to the Recents menu, in a submenu

**Comment by becm** [comment]
If we follow the `git` convention (and refuse to display states that can only be reached with manual file hackery):
There is a 1:1 relation of a `worktree` with an existing local branch.

So it it may (technically) be enough to mark local branches what have a different `worktree` from the current.
And instead of trying to switch the current checkout to a branch that has a different `worktree` (which git will reject) we may instead switch to the `worktree` where it is checked out.

**Comment by mstv** [comment]
> So it it may (technically) be enough to mark local branches what have a different worktree from the current.
And instead of trying to switch the current checkout to a branch that has a different worktree (which git will reject) we may instead switch to the worktree where it is checked out.

👍
- generally when trying to check out a branch, e.g. from RevisionGrid
- with confirmation popup like "Branch "b" is checked out in worktree "path". Switch to this worktree?" 

> There is a 1:1 relation of a worktree with an existing local branch.

It is 1 : 0..1 actually: Commits can be checked out "headless". Then such worktrees would not show up in the left panel. This is not an issue for me because we have the worktrees dialog and I use to switch worktrees using the repo dropdown.

**Comment by drewnoakes** [comment]
PTAL at #12927 which adds worktrees to the left panel.

Since going agentic, I've been using worktrees a lot more and really wanted this feature.
